### PR TITLE
WIP - Updated Code for Bulk Update Deployment Template

### PR DESCRIPTION
**Pull request by kartik-579**
# Description

Updated functions in PipelineConfigRouter, PipelineConfigRestHandler, ChartService & ChartRepository Code files.

Fixes # Bulk Update Deployment template

## Enhancement Issue

- User can Update Deployment template(Apply JSON Patch) in bulk for apps depending upon parameters such as - AppNameIncludes, AppNameExcludes, IsGlobal, EnvId, Patch.

### CODE REVIEW

File: api/restHandler/PipelineConfigRestHandler.go [hunk 1]
```diff
@@ -209,6 +212,37 @@ func NewPipelineRestHandlerImpl(pipelineBuilder pipeline.PipelineBuilder, Logger
 
 const devtron = "DEVTRON"
 
+func (handler PipelineConfigRestHandlerImpl) GetAppNameDeploymentTemplate(w http.ResponseWriter, r *http.Request){
+	decoder := json.NewDecoder(r.Body)
+	var bulkUpdateInput pipeline.BulkUpdateInput
+	err:= decoder.Decode(&bulkUpdateInput)
+	if err!=nil{
```
Comment (pghildiyal) [inline api/restHandler/PipelineConfigRestHandler.go:219]: use formatter, code is not formatted

File: api/restHandler/PipelineConfigRestHandler.go [hunk 2]
```diff
@@ -209,6 +212,37 @@ func NewPipelineRestHandlerImpl(pipelineBuilder pipeline.PipelineBuilder, Logger
 
 const devtron = "DEVTRON"
 
+func (handler PipelineConfigRestHandlerImpl) GetAppNameDeploymentTemplate(w http.ResponseWriter, r *http.Request){
+	decoder := json.NewDecoder(r.Body)
```
Comment (pghildiyal) [inline api/restHandler/PipelineConfigRestHandler.go:216]: auth missing

File: api/restHandler/PipelineConfigRestHandler.go [hunk 3]
```diff
@@ -209,6 +212,37 @@ func NewPipelineRestHandlerImpl(pipelineBuilder pipeline.PipelineBuilder, Logger
...
+	}
+	AppName,err:= handler.chartService.GetBulkAppName(bulkUpdateInput)
+	if err != nil {
+		writeJsonResp(w, err, nil, http.StatusInternalServerError)
+		return
+	}
+	writeJsonResp(w, err, AppName, http.StatusOK)
+}
+func (handler PipelineConfigRestHandlerImpl) BulkUpdateDeploymentTemplate(w http.ResponseWriter, r *http.Request){
+	decoder := json.NewDecoder(r.Body)
```
Comment (pghildiyal) [inline api/restHandler/PipelineConfigRestHandler.go:231]: auth missing

File: internal/sql/repository/chartConfig/ChartRepository.go [hunk 4]
```diff
@@ -22,7 +22,35 @@ import (
 	"github.com/devtron-labs/devtron/internal/sql/repository"
 	"github.com/go-pg/pg"
 )
-
+type App struct{
+	tableName struct{}  		`sql:"app" pg:",discard_unknown_columns"`
```
Comment (pghildiyal) [inline internal/sql/repository/chartConfig/ChartRepository.go:26]: format it

File: pkg/pipeline/ChartService.go [hunk 5]
```diff
@@ -175,6 +187,74 @@ func NewChartServiceImpl(chartRepository chartConfig.ChartRepository,
 	}
 }
 
+
+func(impl ChartServiceImpl) GetBulkAppName(bulkUpdateInput BulkUpdateInput)( []byte, error){
+	BulkAppName := make([]byte,0)
```
Comment (pghildiyal) [inline pkg/pipeline/ChartService.go:192]: format code

Comment (pghildiyal) [inline pkg/pipeline/ChartService.go:192]: why are we returning []bytes and not []string

File: pkg/pipeline/ChartService.go [hunk 6]
```diff
@@ -175,6 +187,74 @@ func NewChartServiceImpl(chartRepository chartConfig.ChartRepository,
...
+	if bulkUpdateInput.IsGlobal {
+		appsGlobal, err := impl.chartRepository.
+			FindBulkAppNameIsGlobal(bulkUpdateInput.AppNameIncludes, bulkUpdateInput.AppNameExcludes)
+		if err != nil {
+		return nil,err
+		}
+		for _,app:=range appsGlobal{
+			BulkAppName = append(BulkAppName,app.AppName+"\n"...)
+		}
+	return BulkAppName,nil
```
Comment (pghildiyal) [inline pkg/pipeline/ChartService.go:202]: should allow both global flag as well as envIds

File: pkg/pipeline/ChartService.go [hunk 7]
```diff
@@ -47,6 +48,14 @@ import (
 	"time"
 )
 
+type BulkUpdateInput struct{
+	AppNameIncludes string  `json:"appNameIncludes"`
+	AppNameExcludes string  `json:"appNameExcludes"`
+	EnvId int `json:"envId"`
```
Comment (pghildiyal) [inline pkg/pipeline/ChartService.go:54]: should take array

File: internal/sql/repository/chartConfig/ChartRepository.go [hunk 8]
```diff
@@ -72,6 +107,74 @@ func NewChartRepository(dbConnection *pg.DB) *ChartRepositoryImpl {
 type ChartRepositoryImpl struct {
 	dbConnection *pg.DB
 }
+func (repositoryImpl ChartRepositoryImpl) FindBulkAppNameIsGlobal(appNameIncludes string,appNameExcludes string) ([]*App,error){
+	var apps []*App
+	err := repositoryImpl.dbConnection.
+		Model(&apps).Join("inner join charts on app.id = app_id").
+		Where("app_name like ? and app_name not like ?","%"+appNameIncludes+"%","%"+appNameExcludes+"%").
```
Comment (pghildiyal) [inline internal/sql/repository/chartConfig/ChartRepository.go:114]: use fmt.Sprintf for concatenating string eg, fmt.Sprintf("%s%s%s", "%", appNameIncludes, "%")

File: internal/sql/repository/chartConfig/ChartRepository.go [hunk 9]
```diff
@@ -72,6 +107,74 @@ func NewChartRepository(dbConnection *pg.DB) *ChartRepositoryImpl {
...
+		Join("inner join chart_env_config_override on charts.id = chart_id").
+		Where("app_name like ? and app_name not like ? and target_environment = ?","%"+appNameIncludes+"%","%"+appNameExcludes+"%",envId).
+		Select()
+	return apps,err
+}
+func (repositoryImpl ChartRepositoryImpl) FindBulkChartsByAppNameSubstring(appNameIncludes string,appNameExcludes string) ([]*Chart,error){
+	var charts []*Chart
+	err := repositoryImpl.dbConnection.
+		Model(&charts).Join("inner join app on app.id=app_id ").
+		Where("app_name like ? and app_name not like ?","%"+appNameIncludes+"%","%"+appNameExcludes+"%").
```
Comment (pghildiyal) [inline internal/sql/repository/chartConfig/ChartRepository.go:132]: use Sprintf

File: internal/sql/repository/chartConfig/ChartRepository.go [hunk 10]
```diff
@@ -72,6 +107,74 @@ func NewChartRepository(dbConnection *pg.DB) *ChartRepositoryImpl {
...
+		Select()
+	return charts,err
+}
+func (repositoryImpl ChartRepositoryImpl) FindBulkChartsEnvByAppNameSubstring(appNameIncludes string,appNameExcludes string,envId int) ([]*ChartEnvConfigOverride,error){
+	var charts []*ChartEnvConfigOverride
+	err := repositoryImpl.dbConnection.
+		Model(&charts).Join("inner join charts on charts.id=chart_id").
+		Join("inner join app on app.id=app_id").
+		Where("app_name like ? and app_name not like ? and target_environment = ?","%"+appNameIncludes+"%","%"+appNameExcludes+"%",envId).
```
Comment (pghildiyal) [inline internal/sql/repository/chartConfig/ChartRepository.go:141]: use Sprintf

File: pkg/pipeline/ChartService.go [hunk 11]
```diff
@@ -175,6 +187,74 @@ func NewChartServiceImpl(chartRepository chartConfig.ChartRepository,
...
+	if err != nil {
+		return nil,err
+	}
+	for _,app:=range appsNotGlobal{
+		BulkAppName = append(BulkAppName,app.AppName+"\n"...)
+	}
+	return BulkAppName,nil
+}
+
+func(impl ChartServiceImpl) BulkUpdateDeploymentTemplate(bulkUpdateInput BulkUpdateInput) ([]byte,error){
```
Comment (pghildiyal) [inline pkg/pipeline/ChartService.go:215]: why returning []bytes and not []string

File: pkg/pipeline/ChartService.go [hunk 12]
```diff
@@ -47,6 +48,19 @@ import (
 	"time"
 )
 
+type BulkUpdateInput struct {
+	AppNameIncludes string `json:"appNameIncludes"`
```
Comment (pghildiyal) [inline pkg/pipeline/ChartService.go:52]: change
AppNameIncludes ->
Includes:
  AppName
and
AppNameExcludes -> 
Excludes:
  AppName

**Comment by kartik-579** [comment]
checked multiple **patch** support, typo in last commit message

**Comment by kartik-579** [comment]
Closing this PR and raising a new one due to change of branch(trial -> batch-update-operation).

**Pull request closed without merging**